File: frontend/src/api/genre.ts
import { apiJson, apiVoid } from "./client";
import { postSseStream } from "./sse";

export type GenreRecommendRequest = {
  targetPlatform: string;
  genderChannel: string;
  preferredGenres: string[];
  avoid: string[];
  writingStrength: string[];
  riskPreference: string;
};

export type SelectedDirection = {
  channel: string;
  genre: string;
  subTags: string[];
  reason: string;
};

export type CandidateRanking = {
  genre: string;
  heatScore: number;
  competitionScore: number;
  payoffDensity: number;
  serializationScore: number;
  originalitySpace: number;
  tokenCostLevel: string;
  finalScore: number;
  recommendReason: string;
  riskNote: string;
};

export type GenreDecisionContract = {
  selectedDirection: SelectedDirection;
  candidateRankings: CandidateRanking[];
  recommendedCoreHook: string;
  riskNotes: string[];
};

export type GenreRecommendResponse = {
  contractId: string;
  contract: GenreDecisionContract;
};

function splitList(s: string): string[] {
  return s
    .split(/[,，]/)
    .map((x) => x.trim())
    .filter(Boolean);
}

export function parseGenreForm(fields: {
  targetPlatform: string;
  genderChannel: string;
  preferredGenresRaw: string;
  avoidRaw: string;
  writingStrengthRaw: string;
  riskPreference: string;
}): GenreRecommendRequest {
  return {
    targetPlatform: fields.targetPlatform.trim(),
    genderChannel: fields.genderChannel.trim(),
    preferredGenres: splitList(fields.preferredGenresRaw),
    avoid: splitList(fields.avoidRaw),
    writingStrength: splitList(fields.writingStrengthRaw),
    riskPreference: fields.riskPreference.trim() || "medium",
  };
}

export function postGenreRecommend(
  projectId: string,
  body: GenreRecommendRequest
): Promise<GenreRecommendResponse> {
  return apiJson<GenreRecommendResponse>(
    `/api/projects/${encodeURIComponent(projectId)}/genre/recommend`,
    {
      method: "POST",
      body: JSON.stringify(body),
    }
  );
}

/** SSE：Java 透传 Writer；`persisted` 事件含落库后的 contractId / contract。 */
export function postGenreRecommendStream(
  projectId: string,
  body: GenreRecommendRequest,
  onFrame: (eventName: string, payload: Record<string, unknown>) => void
): Promise<void> {
  return postSseStream(
    `/api/projects/${encodeURIComponent(projectId)}/genre/recommend/stream`,
    body,
    (eventName, dataJson) => {
      const payload = JSON.parse(dataJson) as Record<string, unknown>;
      onFrame(eventName, payload);
    }
  );
}

export type GenreStoryHookStreamRequest = {
  storyHook: string;
  targetPlatform?: string;
  genderChannel?: string;
  riskPreference?: string;
  /** Skill 定稿后：Writer 单轮唯一题材锁定，无 Scout 多备选 */
  uniqueDirection?: boolean;
};

/** SSE：同一 Writer 流水线，请求体带 storyHook（偏好在后端填空数组）。 */
export function postGenreRecommendFromStoryStream(
  projectId: string,
  body: GenreStoryHookStreamRequest,
  onFrame: (eventName: string, payload: Record<string, unknown>) => void
): Promise<void> {
  return postSseStream(
    `/api/projects/${encodeURIComponent(projectId)}/genre/recommend/from-story/stream`,
    body,
    (eventName, dataJson) => {
      const payload = JSON.parse(dataJson) as Record<string, unknown>;
      onFrame(eventName, payload);
    }
  );
}

export function putGenreSelection(projectId: string, genreContractId: string): Promise<void> {
  return apiVoid(`/api/projects/${encodeURIComponent(projectId)}/genre/selected-contract`, {
    method: "PUT",
    body: JSON.stringify({ genreContractId }),
  });
}

/** 单条题材方案详情（含完整 raw_json）。 */
export type GenreContractDetail = {
  id: string;
  projectId: string;
  createdAt: string;
  source: string;
  storyHookText: string | null;
  rawJson: Record<string, unknown>;
};

export function getGenreContract(projectId: string, contractId: string): Promise<GenreContractDetail> {
  return apiJson<GenreContractDetail>(
    `/api/projects/${encodeURIComponent(projectId)}/genre/${encodeURIComponent(contractId)}`
  );
}

export function putGenreContract(
  projectId: string,
  contractId: string,
  body: { rawJson?: Record<string, unknown>; selectedDirection?: Record<string, unknown> }
): Promise<GenreContractDetail> {
  return apiJson<GenreContractDetail>(
    `/api/projects/${encodeURIComponent(projectId)}/genre/${encodeURIComponent(contractId)}`,
    {
      method: "PUT",
      body: JSON.stringify(body),
    }
  );
}

export function deleteGenreContract(projectId: string, contractId: string): Promise<void> {
  return apiVoid(`/api/projects/${encodeURIComponent(projectId)}/genre/${encodeURIComponent(contractId)}`, {
    method: "DELETE",
  });
}

export type GenreInterviewChatTurn = {
  role: string;
  content: string;
};

export type GenreInterviewResponse = {
  status: string;
  replyToUser: string;
  finalSummary?: string | null;
  coreSettings?: Record<string, unknown> | null;
  persistedNovelSeedContractId?: string | null;
};

export type GenreInterviewRequestBody = {
  chatHistory: GenreInterviewChatTurn[];
  /** 若提供，Writer 注入对应 Skill 全文，模型据此追问细节；可与空 chatHistory 配合自动开场。 */
  writerSkillId?: string | null;
};

/** 路径 B：多轮互动采访（非流式）。可选 writerSkillId 进入 Skill 确认模式。 */
export function postGenreInterview(
  projectId: string,
  chatHistory: GenreInterviewChatTurn[],
  options?: { writerSkillId?: string | null }
): Promise<GenreInterviewResponse> {
  const body: GenreInterviewRequestBody = { chatHistory };
  const sid = options?.writerSkillId?.trim();
  if (sid) {
    body.writerSkillId = sid;
  }
  return apiJson<GenreInterviewResponse>(
    `/api/projects/${encodeURIComponent(projectId)}/genre/interview`,
    {
      method: "POST",
      body: JSON.stringify(body),
    }
  );
}

File: writer-python/app/schemas/genre.py
from __future__ import annotations

from pydantic import BaseModel, ConfigDict, Field, field_validator


def _to_camel(snake: str) -> str:
    parts = snake.split("_")
    return parts[0] + "".join(p.title() for p in parts[1:])


# 请求体：不要用 serialize_by_alias，否则部分 pydantic/FastAPI 版本解析入参异常 → 422
_GENRE_REQUEST_CONFIG = ConfigDict(
    alias_generator=_to_camel,
    populate_by_name=True,
    extra="ignore",
)

# 响应体：序列化为 camelCase，供 Java / 前端读取
_GENRE_RESPONSE_CONFIG = ConfigDict(
    alias_generator=_to_camel,
    populate_by_name=True,
    serialize_by_alias=True,
    extra="ignore",
)


class GenreRecommendRequest(BaseModel):
    """与 Java / 前端一致的 camelCase JSON。"""

    model_config = _GENRE_REQUEST_CONFIG

    target_platform: str = Field(description="目标平台展示名，如 番茄、起点")
    gender_channel: str = Field(description="男频 / 女频 等")
    preferred_genres: list[str] = Field(default_factory=list)
    avoid: list[str] = Field(default_factory=list)
    writing_strength: list[str] = Field(default_factory=list)
    risk_preference: str = Field(default="medium", description="low | medium | high")
    story_hook: str | None = Field(
        default=None,
        description="可选：作者一两句故事线/创意；若填写则各 Agent 须与之对齐",
    )
    project_id: str | None = Field(default=None, description="可选，用于 llm_usage_log")


class SelectedDirection(BaseModel):
    model_config = _GENRE_RESPONSE_CONFIG

    channel: str
    genre: str
    sub_tags: list[str] = Field(default_factory=list)
    reason: str


class CandidateRanking(BaseModel):
    model_config = _GENRE_RESPONSE_CONFIG

    genre: str
    heat_score: float = Field(ge=0, le=10)
    competition_score: float = Field(ge=0, le=10)
    payoff_density: float = Field(ge=0, le=10)
    serialization_score: float = Field(ge=0, le=10)
    originality_space: float = Field(ge=0, le=10)
    token_cost_level: str
    final_score: float = Field(ge=0, le=10)
    recommend_reason: str = Field(description="推荐理由")
    risk_note: str = Field(description="该题材风险说明")


class GenreDecisionContract(BaseModel):
    model_config = _GENRE_RESPONSE_CONFIG

    selected_direction: SelectedDirection
    candidate_rankings: list[CandidateRanking]
    recommended_core_hook: str
    risk_notes: list[str] = Field(default_factory=list)

    @field_validator("candidate_rankings")
    @classmethod
    def three_candidates(cls, v: list[CandidateRanking]) -> list[CandidateRanking]:
        if len(v) != 3:
            raise ValueError("candidateRankings must contain exactly 3 items")
        return v


class ScoutCandidate(BaseModel):
    model_config = _GENRE_RESPONSE_CONFIG

    genre: str
    pitch: str


class ScoutOutput(BaseModel):
    model_config = _GENRE_RESPONSE_CONFIG

    candidates: list[ScoutCandidate]

    @field_validator("candidates")
    @classmethod
    def three(cls, v: list[ScoutCandidate]) -> list[ScoutCandidate]:
        if len(v) != 3:
            raise ValueError("candidates must have length 3")
        return v


class StrategistCandidate(BaseModel):
    model_config = _GENRE_RESPONSE_CONFIG

    genre: str
    sub_tags: list[str]
    hook_line: str
    pitch: str


class StrategistOutput(BaseModel):
    model_config = _GENRE_RESPONSE_CONFIG

    candidates: list[StrategistCandidate]
    recommended_core_hook: str

    @field_validator("candidates")
    @classmethod
    def three(cls, v: list[StrategistCandidate]) -> list[StrategistCandidate]:
        if len(v) != 3:
            raise ValueError("candidates must have length 3")
        return v
